File: GLSF/GLSF/Controllers/DataSchemas.cs
using Microsoft.EntityFrameworkCore;
using System.Collections.Generic;
using System.ComponentModel.DataAnnotations;
using System.ComponentModel.DataAnnotations.Schema;

namespace GLSF.Controllers
{
	public class fishDBContext : DbContext
	{
		public fishDBContext(DbContextOptions<fishDBContext> options) : base(options)
		{

		}

		protected override void OnModelCreating(ModelBuilder builder) {
			builder.Entity<Boat>().HasKey(table => new
			{
				table.Id,
				table.TournamentId
			});

			builder.Entity<Station>().HasKey(table => new
			{
				table.Id,
				table.TournamentId
			});

			builder.Entity<Member>().HasKey(table => new
			{
				table.Id,
				table.BoatId,
				table.TournamentId
			});

			builder.Entity<Member>().HasOne<Boat>().WithOne().HasForeignKey<Member>(x => new { x.BoatId, x.TournamentId});
			builder.Entity<Fish>().HasOne<Member>().WithOne().HasForeignKey<Fish>(x => new { x.MemberId, x.TournamentId, x.BoatId });
			builder.Entity<Fish>().HasOne<Station>().WithOne().HasForeignKey<Fish>(x => new { x.StationNumber, x.TournamentId });
			builder.Entity<Fish>().HasOne<Boat>().WithOne().HasForeignKey<Fish>(x => new { x.BoatId, x.TournamentId });
			builder.Entity<Member>().HasOne<Tournament>().WithOne().HasForeignKey<Member>(x => x.TournamentId);
			builder.Entity<Fish>().HasOne<Tournament>().WithOne().HasForeignKey<Fish>(x => x.TournamentId );
			builder.Entity<Boat>().HasOne<Tournament>().WithOne().HasForeignKey<Boat>(x => x.TournamentId);
			builder.Entity<Station>().HasOne<Tournament>().WithOne().HasForeignKey<Station>(x => x.TournamentId);

		}
		public DbSet<Fish> Fishes { get; set; }
		public DbSet<Tournament> Tournaments { get; set; }
		public DbSet<Boat> Boats { get; set; }
		public DbSet<Member> Members { get; set; }
		public DbSet<Station> Stations { get; set; }
	}

	public class Fish
	{
		public double Weight { set; get; }
		public double Length { set; get; }
		public string Species { set; get; }
		public string? Image { set; get; }
		public string Date { set; get; }
		public string? SampleNumber { set; get; }
		public bool HasTag { set; get; }
		public string? Port { set; get; }
		public bool IsValid { set; get; }
		public int StationNumber{ get; set; }
		public int MemberId { get; set; }
		public int TournamentId { get; set; }
		public int BoatId { get; set; }
		[Key]
		public int? Id { get; set; }
	}

	public class Tournament
	{
		public string StartDate { get; set; }
		public string EndDate { get; set; }
		public string Name { get; set; }
		public string Location { get; set; }
		[Key]
		public int? Id { get; set; }
	}

	public class Boat
	{
		public string Name { get; set; }
		public double Length { get; set; }
		[Key, Column(Order = 0)]
		public int Id { get; set; }
		[Key, Column(Order = 1)]
		public int TournamentId { get; set; }

	}

	public class Member
	{
		public string Name { get; set; }
		public int Age { get; set; }
		public bool IsCaptain { get; set; }
		public bool IsJunior { get; set; }
		[Key, Column(Order = 0)]
		public int Id { get; set; }
		[Key, Column(Order = 1)]
		public int BoatId { get; set; }
		[Key, Column(Order = 2)]
		public int TournamentId { get; set; }
	}

	public class Group
	{
		public Boat Boat { get; set; }
		public List<Member> Members { get; set; }
	}

	public class Station
	{
		public string Port { get; set; }
		[Key, Column(Order = 0)]
		public int Id { get; set; }
		[Key, Column(Order = 1)]
		public int TournamentId { get; set; }

	}
}

File: GLSF/GLSF/ClientApp/src/app/models/dataSchemas.ts
//Station class/data model for database and server
export class Station {
	Port: string;
	Id: number;
	TournamentId: any;
}

//Boat class/data model for database and server
export class Boat {
	Name: string;
	Length: number;
	CheckedIn: boolean;
	Id: number;
	TournamentId: any;
}

//Member class/data model for database and server
export class Member {
	Name: string;
	Age: any;
	IsCaptain: boolean;
	IsJunior: boolean;
	Id: number;
	BoatId: number;
	TournamentId: any;
}

//Group class/data model for server
//Used to easily send a boat and the attached members
export class Group {
	Boat: Boat;
	Members: Array<Member>;
}


//Tournament class/data model for server and database
export class Tournament {
	StartDate: string;
	EndDate: string;
	Name: string;
	Location: string;
	Id: any;
}

//Time class for tournament hour/minute entry fields
//Makes it simple to add specified time to a date
export class Time{
	Hours: number;
	Minutes: number;
}


//Fish class/data model for server and database
export class Fish {
	Weight: number;
	Length: number;
	Species: string;
	Image: any;
	Date: string;
	SampleNumber: any;
	HasTag: boolean;
	Port: string;
	IsValid: boolean;
	NoClips: boolean;
	FinsClipped: string;
	StationNumber: number;
	MemberId: number;
	TournamentId: any;
	BoatId: number;
	Id: any;

	//Names of fish in lake michigan, expand if needed
	public static fishes = [
		"Atlantic Salmon",
		"Brown Trout",
		"Chinook Salmon",
		"Coho Salmon",
		"Lake Trout",
		"Rainbow Trout",
		"Sauger",
		"Yellow Perch",
		"Walleye",
    "Other", 
	];

	//Lbs of each specified fish in a dictionary (Maxes)
	public static maxWeights = {
		"Chinook Salmon": 130,
		"Coho Salmon": 40,
		"Atlantic Salmon": 110,
		"Brown Trout": 60,
		"Lake Trout": 110,
		"Rainbow Trout": 40,
		"Yellow Perch": 10,
		"Walleye": 35,
		"Sauger": 20,
    "Other": 200,
	};

	//Inches of each specified fish in a dictionary (Maxes)
	public static maxLengths = {
		"Chinook Salmon": 60,
		"Coho Salmon": 60,
		"Atlantic Salmon": 70,
		"Brown Trout": 60,
		"Lake Trout": 60,
		"Rainbow Trout": 60,
		"Yellow Perch": 15,
		"Walleye": 60,
		"Sauger": 40,
    "Other": 80,
    };
}

//User class/data model for database and server
export class User {
	Id: any;
	Username: string;
	Password: string;
	FirstName: string;
	LastName: string;
	Token: string;
	AccessLevel: number;
}
